src/io/eval.rs

```diff
@@ -140,158 +97,75 @@ where
-                let r = public_data.rs[*prev_num as usize].clone();
-                cur_s = cur_s * r;
-            }
-            let new_s = cur_s.clone() * &public_data.rs[*num as usize];
-            let b_next_bit = bs[level + 1][*num as usize].clone();
-            let expected_q = cur_s.concat_columns(&[&new_s]) * &b_next_bit;
-            assert_eq!(q, expected_q);
-            let expected_p = new_s.concat_columns(&[&new_s]) * &bs[level + 1][level_size];
-            assert_eq!(p, expected_p);
-            let expcted_new_encode = {
-                let dim = params.ring_dimension() as usize;
-                let one = <M::P as Poly>::const_one(&params);
-                let gadget_d1 = M::gadget_matrix(&params, d1);
-                let inserted_poly_gadget = {
-                    let mut polys = vec![];
-                    polys.push(one);
-                    let mut coeffs = vec![];
-                    for bit in inputs[0..(level_width * (level + 1))].iter() {
-                        if *bit {
-                            coeffs.push(<M::P as Poly>::Elem::one(&params.modulus()));
-                        } else {
-                            coeffs.push(<M::P as Poly>::Elem::zero(&params.modulus()));
-                        }
+            let r = &public_data.rs[*num as usize];
+            s_cur = s_cur * r;
+            let bits_done = level_width * level;
+            let dim = params.ring_dimension() as usize;
+            let mut polys: Vec<M::P> = vec![<M::P as Poly>::const_one(&params)];
+            let mut coeffs = inputs[..bits_done]
+                .iter()
+                .map(|&b| {
+                    if b {
+                        <M::P as Poly>::Elem::one(&params.modulus())
+                    } else {
+                        <M::P as Poly>::Elem::zero(&params.modulus())
                     }
-                    for _ in 0..(obf_params.input_size - level_width * (level + 1)) {
-                        coeffs.push(<M::P as Poly>::Elem::zero(&params.modulus()));
-                    }
-                    let input_polys = coeffs
-                        .chunks(dim)
-                        .map(|coeffs| M::P::from_coeffs(&params, coeffs))
-                        .collect_vec();
-                    polys.extend(input_polys);
-                    polys.push(minus_t_bar.clone());
-                    M::from_poly_vec_row(&params, polys).tensor(&gadget_d1)
-                };
-                let pubkey = pub_key_cur[0].concat_matrix(&pub_key_cur[1..]);
-                new_s * (pubkey - inserted_poly_gadget)
-            };
-            assert_eq!(new_encode_vec, expcted_new_encode);
+                })
+                .collect::<Vec<_>>();
+            coeffs.extend(std::iter::repeat_n(
+                <M::P as Poly>::Elem::zero(&params.modulus()),
+                obf_params.input_size - bits_done,
+            ));
+            polys.extend(coeffs.chunks(dim).map(|c| M::P::from_coeffs(&params, c)));
```

Comment:
We are repeating this logic of building polynomials from `inputs[..bits_done]` at least three times in the `evaluate` function. Can we maybe wrap it into a utils?